[G-Y-Chan] iP

G-Y-Chan opened the pull request:
### SecondMind

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

SecondMind frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](url).
2. double-click it.
3. add you tasks.
4. let it manage your tasks for you 🥵

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

<hr />

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

G-Y-Chan committed:
Rename java class used for chatbot

G-Y-Chan committed:
Replace logo String with a placeholder logo

G-Y-Chan committed:
Create function to print line separators between messages

G-Y-Chan committed:
Create function to greet user

G-Y-Chan committed:
Create function to exit program

G-Y-Chan committed:
Replace old code for greeting with a function call to greetUser

G-Y-Chan committed:
Create function to echo commands entered by user

G-Y-Chan committed:
Create String array to store the list of tasks

G-Y-Chan committed:
Create function to add tasks to the taskList array

G-Y-Chan committed:
Replace code that echoes command with a function call to addToTaskList()

G-Y-Chan committed:
Add function to print out task list

G-Y-Chan committed:
Add logic to call addToTaskList function

G-Y-Chan committed:
Create Task class to encapsulate data related to each class within an object

G-Y-Chan committed:
Create functions to mark a task as done or undone

G-Y-Chan committed:
Override toString() method to provide a clearer representation of the Task object

G-Y-Chan committed:
Create function to get description of task

G-Y-Chan committed:
Replace String representations of tasks with Task objects

G-Y-Chan committed:
Move logic for processing input to another function

G-Y-Chan committed:
Create functions to mark/unmark tasks and notify the user

G-Y-Chan committed:
Add logic for processing mark/unmark commands

G-Y-Chan committed:
Add function call to print missing line separator

G-Y-Chan committed:
Create new sublass of Task for tasks without date/time

G-Y-Chan committed:
Create function to support creation of different types of tasks

G-Y-Chan committed:
Modify function to support creation of DeadlineTask objects

G-Y-Chan committed:
Create new subclass of Task for tasks with a deadline

G-Y-Chan committed:
Create new subclass of Task for events

G-Y-Chan committed:
Modify function to support creation of EventTask objects

G-Y-Chan committed:
Modify function to notify user about the number of tasks in the list

G-Y-Chan committed:
Updated respective files for TextUI testing

G-Y-Chan committed:
Create an exception to be thrown when description of todo event is empty

G-Y-Chan committed:
Create an exception to be thrown when an unknown command is entered

G-Y-Chan committed:
Create exception to be thrown when an empty command is detected

G-Y-Chan committed:
Split logic within createEvent into different functions for different event types

G-Y-Chan committed:
Create function that improves readability of error messages

G-Y-Chan committed:
Create exception to deal with invalid task numbers

G-Y-Chan committed:
Modify function to deal with invalid task numbers

G-Y-Chan committed:
Use an Arraylist to keep track of the task list instead

G-Y-Chan committed:
Create function to remove a task from the task list

G-Y-Chan committed:
Modify function to support delete command

G-Y-Chan committed:
Create empty text file to store list of tasks

G-Y-Chan committed:
Store file path as a constant String variable

G-Y-Chan committed:
Create function to convert text from data file into a Task object

G-Y-Chan committed:
Create function to load task list from .txt file

G-Y-Chan committed:
Modify main() function to load task list from .txt file

G-Y-Chan committed:
Create function that appends given data to end of data file

G-Y-Chan committed:
Fix bug where function was overwriting entire data file

G-Y-Chan committed:
Modify function to save new todo task to data file

G-Y-Chan committed:
Modify function to save new deadline task to data file

G-Y-Chan committed:
Simplify processing of the input String

G-Y-Chan committed:
Modify function to save new event task to data file

G-Y-Chan committed:
Fix bug where completed status of task was not reflected

G-Y-Chan committed:
Modify function to delete selected task from data file

G-Y-Chan committed:
Create function that updates boolean flag of given task in .txt file

G-Y-Chan committed:
Modify function to update boolean flag of task in data file

G-Y-Chan committed:
Modify function to update boolean flag of task in data file

G-Y-Chan committed:
Modify function to create new file when FileNotFound exceptions are caught

G-Y-Chan committed:
Merge branch 'branch-Level-7'

G-Y-Chan committed:
Create function to convert String format of dates to LocalDateTime objects

G-Y-Chan committed:
Create function to convert LocalDateTime object to desired string format

G-Y-Chan committed:
Fix typo in function name

G-Y-Chan committed:
Combine two functions into one for better readability

G-Y-Chan committed:
Modify function to call on formatDateTime()

G-Y-Chan committed:
Modify function to catch DateTimeParseExceptions and inform user on correct format

G-Y-Chan committed:
Modify format of String to be returned

G-Y-Chan committed:
Modify function to call on formatDateTime()

G-Y-Chan committed:
Merge branch 'branch-Level-8'

G-Y-Chan committed:
Refactor code to extract out closely related code as classes

G-Y-Chan committed:
Merge branch 'branch-A-MoreOOP'

G-Y-Chan committed:
Move code related to UI into a new file

G-Y-Chan committed:
Move any code related to printing to UI class

G-Y-Chan committed:
Merge branch 'branch-A-MoreOOP'

G-Y-Chan committed:
Remove unnecessary import statements

G-Y-Chan committed:
Move all java files into secondmind package

G-Y-Chan committed:
Merge branch 'branch-A-Packages'

G-Y-Chan committed:
Remove code previously used for testing

G-Y-Chan committed:
Merge branch 'add-gradle-support'

G-Y-Chan committed:
Write test cases for TaskList class

G-Y-Chan committed:
Fix bug for storage representation of EventTask objects

G-Y-Chan committed:
Write test cases for Parser class

G-Y-Chan committed:
Merge branch 'branch-A-JUnit'

G-Y-Chan committed:
Modify file path of data file

yadobler commented on src/main/java/secondmind/DeadlineTask.java:
```diff
@@ -0,0 +1,27 @@
+package secondmind;
+
+public class DeadlineTask extends Task {
```
No JavaDoc? 😢

yadobler replied:
Ok Nice, do resolve this when availalbe!

yadobler commented on src/main/java/secondmind/Parser.java:
```diff
@@ -0,0 +1,39 @@
+    public Parser() {}
+
+    public String[] processInput(String input) throws NumberFormatException {
+        String[] newInput = input.split(" ");
+        String[] ignoreInstruction = new String[] {"ignore"};
```
Perhaps the `newInput` and `ignoreInstruction` can be pluralised? Even though there's one item, it's still a list that can hold multiple items.

yadobler commented on src/main/java/secondmind/SecondMind.java:
```diff
@@ -0,0 +1,151 @@
+package secondmind;
+
+import java.util.ArrayList;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.time.format.DateTimeParseException;
```
Imports are not in lexical order, maybe sort by alphabetical order?

yadobler commented on src/main/java/secondmind/SecondMind.java:
```diff
@@ -0,0 +1,151 @@
+        while (true) {
+            String input = ui.getInput();
+            try {
+                String[] instruction = parser.processInput(input);
```
Why is `instruction` singular here?

yadobler commented on src/main/java/secondmind/Storage.java:
```diff
@@ -0,0 +1,122 @@
+            String currentLine = s.nextLine();
+            if (lineNumber == taskNumber) {
+                StringBuilder sb = new StringBuilder();
+                //Add task type and "|" to stringbuilder
```
1. Definitely a great thing adding comments to explain lines here and there!
2. Coding conventions suggest a space after `//` and `*` if doing code blocks

yadobler commented on src/main/java/secondmind/Storage.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    private Task textToTask(String text) {
+        String[] taskInfo = text.split("\\|");
```
Why is `taskInfo` singular here?

yadobler commented on src/main/java/secondmind/TaskList.java:
```diff
@@ -0,0 +1,117 @@
+package secondmind;
+
+import java.util.ArrayList;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeParseException;
+import java.time.format.DateTimeFormatter;
```
Imports can be in lexical order if possible

yadobler commented on src/main/java/secondmind/TaskList.java:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    public void delete(int taskNumber) throws InvalidTaskNumberException {
```
1. I love that you named all the methods appropriately, it leaves very little room for ambiguity. 😄 
2. Might `delete(int)` be confused for deleting the entire list instead of a task? (might be clearer to use `deleteTask`)

yadobler commented on src/main/java/secondmind/UI.java:
```diff
@@ -0,0 +1,46 @@
+        printLineSeparator();
+    }
+
+    public void output(String text) {
+        printLineSeparator();
+        System.out.println(text);
+        printLineSeparator();
+    }
```
Would naming the method `output` be too ambiguous, over something like `outputString`?

yadobler commented on src/test/java/secondmind/ParserTest.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.ArrayList;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Static statements usually should be the first in the list of imports

yadobler replied:
Can this wildcard import be expanded to single imports?

yadobler reviewed (commented):
1. Javadoc is missing 😢. This affects how much I can comment on your coding style.
2. Imports can be reorganised, avoiding wildcards.
3. Containers / Collections (ie Arrays and ArrayLists) should have pluralised names.
5. But overall great naming convention. 😄 

Do see how you can resolve these. I based them off of [the java coding standards](https://se-education.org/guides/conventions/java/intermediate.html), as well as the [Intellij configuration for coding styles.](https://se-education.org/guides/tutorials/intellijCodeStyle.html)

G-Y-Chan committed:
Add JavaDoc comments to SecondMind class

G-Y-Chan committed:
Add JavaDoc comments to Storage class

G-Y-Chan committed:
Add JavaDoc comments to TaskList class

G-Y-Chan committed:
Add JavaDoc comments to Parser class

G-Y-Chan committed:
Add JavaDoc comments to UI class

G-Y-Chan committed:
Add JavaDoc comments to Task class

G-Y-Chan committed:
Add JavaDoc comments to ToDoTask class

G-Y-Chan committed:
Add JavaDoc comments to DeadlineTask class

G-Y-Chan committed:
Add JavaDoc comments to EventTask class

G-Y-Chan committed:
Create method that returns an ArrayList of tasks that contains the input string

G-Y-Chan committed:
Modify method to support find commands

G-Y-Chan committed:
Merge branch 'branch-A-JavaDoc'

G-Y-Chan committed:
Resolve merge conflict

G-Y-Chan committed:
Add missing JavaDoc comments

yadobler reviewed (commented):
LGTM!

Do mark the review comments as resolved when done.

G-Y-Chan committed:
Add user and second mind images for GUI

G-Y-Chan committed:
Add fxml file for DialogBox

G-Y-Chan committed:
Add fxml file for main window in GUI

G-Y-Chan committed:
Refactor code to include GUI functionality

G-Y-Chan committed:
Merge branch 'branch-Level-10'

G-Y-Chan committed:
Ensure processInput() method does not return null instructions

G-Y-Chan committed:
Ensure task description is not null

G-Y-Chan committed:
Ensure task deadline is not null

G-Y-Chan committed:
Ensure eventStart and eventEnd are not null

G-Y-Chan committed:
Merge branch 'branch-A-Assertions'

G-Y-Chan committed:
Create named constant for exit instruction string

G-Y-Chan committed:
Create method to encapsulate code related to instruction retrieval

G-Y-Chan committed:
Create method to encapsulate code related to instruction execution

G-Y-Chan committed:
Encapsulate code related to executing "mark" instructions

G-Y-Chan committed:
Encapsulate code related to retrival of task number from instructions

G-Y-Chan committed:
Encapsulate code related to modifying status of tasks

G-Y-Chan committed:
Encapsulate code related to creating completion message

G-Y-Chan committed:
Encapsulate code related to formatting exception message

G-Y-Chan committed:
Modify methods to support "unmark" instructions too

G-Y-Chan committed:
Encapsulate code related to unmarking tasks

G-Y-Chan committed:
Fix bug related to modifying task status

G-Y-Chan committed:
Modify execute() method

There are multiple levels of abstraction found within a code fragment
in the execute() method, which is related to the deletion of a task
from the task list and storage.

Extracting the relevant code and creating multiple functions to handle
different aspects of the deletion function will allow for greater
readability of the code, making it easier to understand and debug,
if necessary.

G-Y-Chan committed:
Modify method name to better reflect its functionality

G-Y-Chan committed:
Encapsulate code related to listing out tasks

G-Y-Chan committed:
Modify execute() function

There are multiple levels of abstraction found within a code fragment
in the execute() method, which is related to finding tasks with
descriptions containing the input match string from user.

Extracting the relevant code and creating multiple functions to handle
different aspects of the matching function will allow for greater
readability of the code, making it easier to understand and debug,
if necessary.

G-Y-Chan committed:
Refactor SecondMind and TaskList classes

There are multiple levels of abstraction found within code fragments
in SecondMind and TaskList classes, which are related to the cretaion
of different Task Objects.

Extracting the relevant code and creating multiple functions to handle
different aspects of the task creation and adding tasks to the TaskList
will allow for greater readability of the code, making it easier to
understand and debug, if necessary.

G-Y-Chan committed:
Fix bug related to wrong number of tasks displayed after deletion

G-Y-Chan committed:
Replace if-else block with switch block for better readability

G-Y-Chan committed:
Refactor Storage class

There are multiple levels of abstraction found within different code
fragments in the Storage class.

Extracting the relevant code and creating multiple functions to handle
different aspects of the Storage class will allow for greater
readability of the code, making it easier to understand and debug,
if necessary.

G-Y-Chan committed:
Remove "magic" strings

G-Y-Chan committed:
Merge pull request #2 from G-Y-Chan/branch-A-CodeQuality

Branch a code quality

G-Y-Chan committed:
Override equals() method to compare equality with other Task objects

G-Y-Chan committed:
Create method to check if duplicate task exists

G-Y-Chan committed:
Modify method to support checking for duplicate task

G-Y-Chan committed:
Merge branch 'branch-C-DetectDuplicates'

G-Y-Chan committed:
Fix bug causing leading whitespace in task descriptions

G-Y-Chan committed:
Update test cases to reflect new behaviour of TaskList class

G-Y-Chan committed:
Sort import statements lexicographically

G-Y-Chan committed:
Sort import statements lexicographically

G-Y-Chan committed:
Modify exception message string for better clarity of instruction

G-Y-Chan committed:
Remove redundant try catch block

G-Y-Chan committed:
Add new exception class to be thrown if a date entered has passed

G-Y-Chan committed:
Refactor multiple classes

There are multiple classes that contain methods which throw
invalid task number exceptions. However, the methods used to deal with
such exceptions differ throughout the different classes, and there are
instances of using try catch blocks with this exception to control the
flow of logic. The error message for InvalidTaskNumberException is also
specialised to numbers that are non-positive or larger than the
number of tasks in the task list.

Refactoring the relavant classes and removing redundant try catch
blocks allows for better readability and hence easier debugging.
Generalising the error message for InvalidTaskNumberException also
helps with providing error messages that are more relevant and easier
understand.

G-Y-Chan committed:
Fix bug related to "find" instruction

G-Y-Chan committed:
Generalise exception for all task types

G-Y-Chan committed:
Modify program to handle invalid command formats

G-Y-Chan committed:
Remove code comented out previously

G-Y-Chan committed:
Remove redundant else block

G-Y-Chan committed:
Modify error message shown

G-Y-Chan committed:
Modify method name

G-Y-Chan committed:
Update README.md

Skeletal user guide with all non-trivial commands.

G-Y-Chan committed:
Replace duke with SecondMind

G-Y-Chan committed:
Modify method to display task numbers in list

G-Y-Chan committed:
Merge branch 'master' of https://github.com/G-Y-Chan/ip

G-Y-Chan committed:
Update README.md

Add outputs for example commands

G-Y-Chan committed:
Add product screenshot via upload

G-Y-Chan committed:
Update README.md

Add link to product screenshot

G-Y-Chan committed:
Modify method to display task numbers beside task descriptions

G-Y-Chan committed:
Rename class for invalid command format exceptions

G-Y-Chan committed:
Create border around responses

G-Y-Chan committed:
Increase window size to minimise wrapping of longer task descriptions

G-Y-Chan committed:
Refactor multiple classes

There are runtime errors that are caused by invalid command formats
such as delete, mark and unmark commands being called without
specifying the task number.

Creating a new exception will allow the user to know that the command
format being used currently is incorrect.

G-Y-Chan committed:
Modify and add test cases to reflect new behaviour of Parser class

G-Y-Chan committed:
Merge branch 'master' of https://github.com/G-Y-Chan/ip

G-Y-Chan committed:
Remove exception from heading that is never thrown

G-Y-Chan committed:
Rename product screenshot file

G-Y-Chan committed:
Rename file path for product screenshot
